lambdas/access_counts/index.py

```diff
@@ -0,0 +1,200 @@
+"""
+Lambda function that runs Athena queries over CloudTrail logs and .quilt/named_packages/
+and creates summaries of object and package access events.
+"""
+
+import os
+import textwrap
+import time
+
+import boto3
+
+ATHENA_DATABASE = os.environ['ATHENA_DATABASE']
+# Bucket where query results will be stored.
+QUERY_RESULT_BUCKET = os.environ['QUERY_RESULT_BUCKET']
+# A temporary directory where Athena query results will be written.
+QUERY_TEMP_DIR = os.environ['QUERY_TEMP_DIR']
+# Directory where the summary files will be stored.
+ACCESS_COUNTS_OUTPUT_DIR = os.environ['ACCESS_COUNTS_OUTPUT_DIR']
+
+
+def sql_escape(s):
+    return s.replace("'", "''")
+
+
+DROP_OBJECT_ACCESS_LOG = """DROP TABLE IF EXISTS object_access_log"""
+DROP_PACKAGE_HASHES = """DROP TABLE IF EXISTS package_hashes"""
+
+CREATE_OBJECT_ACCESS_LOG = textwrap.dedent(f"""\
+    CREATE TABLE object_access_log
+    WITH (
+        format = 'Parquet',
+        parquet_compression = 'SNAPPY',
+        external_location = 's3://{sql_escape(QUERY_RESULT_BUCKET)}/{sql_escape(QUERY_TEMP_DIR)}/object_access_log/'
+    )
+    AS
+    SELECT eventname, date_format(from_iso8601_timestamp(eventtime), '%Y-%m-%d') AS date, bucket, key
+    FROM (
+        SELECT
+            eventname,
+            eventtime,
+            json_extract_scalar(requestparameters, '$.bucketName') AS bucket,
+            json_extract_scalar(requestparameters, '$.key') AS key
+        FROM cloudtrail
+    )
+    WHERE bucket IS NOT NULL AND key IS NOT NULL
```

Comment:
is this WHERE clause insurance or are there cases when bucket or key can be null? if the latter, worth a comment.

Reply:
Well, CloudTrail can contain all kinds of events - user logged in, etc. - not even S3 related. (But even S3 events like ListBucket don't have a key.)